File: 05_Check_NCBI.py


#######################
# Aim: download species from NBCI to create a Reference Database, if species is not represented find sequences from same genus
# have to be added check if fasta file is empty
#######################

import os
import sys
import re
import numpy
from Bio import SeqIO
from Bio import Entrez
from Bio.Seq import Seq
from datetime import datetime
import shutil
import time
from io import StringIO


# Set a true global timeout for all Entrez calls
import urllib
# Correct way: return raw binary stream — Entrez handles parsing
def _patched_open(*args, **kwargs):
    kwargs['timeout'] = 30  # or your TIMEOUT
    resp = urllib.request.urlopen(*args, **kwargs)
    content_type = resp.info().get_content_charset() or "utf-8"
    # wrap as text if retmode=text
    return io.TextIOWrapper(resp, encoding=content_type)




# define NCBI timeout time 
TIMEOUT=30
CONFIG_file="00_login_data.ini"


## read in config file with email
import configparser
config = configparser.ConfigParser()
with open(CONFIG_file) as f:
    config.read_string("[DEFAULT]\n" + f.read())

Entrez.email = config["DEFAULT"]["Entrez_email"]

###### extract date
date = datetime.now()
date = date.strftime("%d_%b_%Y")
date = str(date)


### Steps
FLAG_search_spec=True ### Step 1 - if species list should be 
FLAG_search_gen=True ### Step 2 - check if other species of the same genus are represented (depends on Step 1)

#ncbi search parameter 
search=" AND (18S OR small ribosomal) NOT (chloroplast[filter] OR mitochondrion[filter]) AND (350[SLEN] : 5000[SLEN])"
ncbi_db="nucleotide"

#input files are:genus,species_name\n (have to contain a header)
Input_folder="01_intermediate_results/"
file_missing = Input_folder+"4.1_Missing_Species.csv" ## species missing in database (Step 1)
file_missing_but_genus=Input_folder+"4.2_Present_Species.csv"  ## complete species list in pr2 for genera which still miss species(step 1 & 2)
file_missing_but_genus_new = Input_folder+"4.2_Present_Species_completed_withNCBI.csv" 
shutil.copy(file_missing_but_genus, file_missing_but_genus_new )


#output
NCBI_folder="02_NCBI_"+ date +"/" # where to store the downloaded files
NCBI_results_folder="02_NCBI_"+ date +"_results/" # where to store the downloaded files
output_prefix="05_Missing_PR2" # prefix of the output files

#######################################################################################################################
################################### PLEASE DON'T CHANGE ANYTHING FROM HERE ON ########################################
#######################################################################################################################

################ VARIABLES
#output
file_still_missing= NCBI_results_folder + output_prefix +"_Species_and_missing_NCBI__"+ date +".csv"
file_missing_genera= NCBI_results_folder +output_prefix +"_incomplete_genera_despite_NCBI__"+ date +".csv"
file_missing_genera2= NCBI_results_folder +output_prefix +"_MISSING_genera_despite_NCBI__"+ date +".csv"
file_asseccion_info= NCBI_results_folder +output_prefix +"_Downloaded_species_NCBI_info__"+ date +".tsv"
file_LOG= NCBI_results_folder + output_prefix + "LOG__" + date +".txt"
file_LOG2=NCBI_results_folder + output_prefix + "LOG__" + date +"_successfull_genera.txt"

if not os.path.exists(NCBI_folder):
   os.makedirs(NCBI_folder)

if not os.path.exists(NCBI_results_folder):
   os.makedirs(NCBI_results_folder)
################################################################################################################

# set so Entrez is underlying the try if there is long response or similar
#Entrez._open = _patched_open

#1.) check if species exists in ncbi and download fasta
if FLAG_search_spec:
    with open(file_missing, "r") as f, open(file_still_missing, "w") as f_new, open(file_missing_but_genus_new, "a") as f_new2,open(file_LOG, "a") as f_log, open(file_asseccion_info, "w") as f_assec:
        lines = f.readlines()[1:]
        for line in lines:
                line = line.strip() # remove whtie space and newline character from beginning and the end
                line = line.replace("\"","")
                line_information = line.split(",")
                print(line_information[1])
                new_search = line_information[1]+"[ORGANISM]"+ search
                f_log.write(new_search+ "\n") 
                attempts = 0
                while attempts < 3:
                    try:
                        print(attempts+1)
                        # check if species exist on ncbi
                        handle = Entrez.esearch(db=ncbi_db, retmax=10,term= new_search, timeout=TIMEOUT)
                        record = Entrez.read(handle)
                        handle.close()

                        # check if ncbi request was positive, if not continue
                        if len(record['IdList']) == 0:
                            f_new.write(line+ "\n") 
                            break

                        print("try to download... " )
                        id_list = list(record['IdList'])
                        out_file = NCBI_folder + line_information[1].replace(" ", "_") + "__NCBI.fasta"

                        # Download FASTA, should now catch the try function 
                        with Entrez.efetch(db=ncbi_db, id=id_list, rettype="fasta", retmode="text") as handle_fasta:
                            fasta_text = handle_fasta.read()  # Already str, no need to decode
                            all_records = SeqIO.parse(StringIO(fasta_text), 'fasta')
                            count = SeqIO.write(all_records, out_file, "fasta-2line")

                        # Check if file was created
                        if os.path.isfile(out_file):
                            f_new2.write(line_information[0] + "," + line_information[1] + "\n")

                            # Download GenBank and extract taxonomy info
                            with Entrez.efetch(db=ncbi_db, id=id_list, rettype="genbank", retmode="text") as handle_genbank:
                                genbank_text = handle_genbank.read()  # Already str, no need to decode
                                for record in SeqIO.parse(StringIO(genbank_text), 'genbank'):
                                    f_assec.write(
                                        line_information[1] + "\t" +
                                        record.annotations.get('organism', 'N/A') + "\t" +
                                        record.id + "\t" +
                                        out_file + "\t" +
                                        ";".join(record.annotations.get('taxonomy', [])) + "\n"
                                    )

                        break # success
                    except Exception as e:
                        print(f"Something went wrong with line '{line.strip()}': {e}")
                        f_log.write(line + f"\tproblem: {type(e).__name__} - {e}\n")
                        attempts += 1



#sys.exit()

if FLAG_search_gen:
    #2.) make list of genera where some species are missing but some already exists
    genera_dict =dict()
    with open(file_missing_but_genus_new, "r") as f:
        lines = f.readlines()[1:]
        for line in lines:
            line = line.strip() # remove whtie space and newline character from beginning and the end
            line = line.replace("\"","")
            line_information = line.split(",")
            if line_information[0] in genera_dict:
                genera_dict[line_information[0]].append(line_information[1])  
            else:
                genera_dict[line_information[0]] = [line_information[1]]


    #3.) make list of genera to search for
    genera_search = []
    with open(file_still_missing, "r") as f:
            lines = f.readlines()
            for line in lines:
                line = line.strip() # remove whtie space and newline character from beginning and the end
                line = line.replace("\"","")
                line_information = line.split(",")
                if line_information[0] not in genera_search:
                    genera_search.append(line_information[0] )

    # 4.) check genera on ncbi and download
    with open(file_missing_genera, "w") as f_incomp, open(file_missing_genera2, "w") as f_miss, open(file_LOG, "a") as f_log, open(file_asseccion_info, "a") as f_assec, open(file_LOG2, "w") as f_log2:
        for genus in genera_search:
            print("")
            print(genus)
            #create search link
            new_search=""
            if genus in genera_dict:
                new_search = genus + "[ORGANISM] NOT (" +  "[ORGANISM] OR ".join(genera_dict[genus]) + "[ORGANISM]) " + search
                #print(new_search)
                #sys.exit()
            else:
                new_search = genus+ "[ORGANISM] " +  search

            attempts = 0
            while attempts < 3:
                try:
                    # search if genus is represented at ncbi except for already downloaded species
                    handle = Entrez.esearch(db=ncbi_db, retmax=50,term= new_search)
                    record = Entrez.read(handle)

                    #check if feedback is positive ore not
                    if len(record['IdList']) == 0:
                        print("Genus unknown: \t"+genus)
                        if genus in genera_dict:
                            f_miss.write(genus+","+ "&".join(genera_dict[genus])+ new_search +"\n") 
                        else:
                            f_incomp.write(genus+","+ new_search +"\n") 
                    else:
                        f_log2.write(genus+","+ new_search +"\n") 
                        #download genbank entries to get further organism info
                        handle = Entrez.efetch(db=ncbi_db, id=list(record['IdList']), rettype="genbank", retmode="text")
                        out_file=NCBI_folder + genus +"__NCBI.fasta"

                        #download fasta entries to save fasta file (otherwise empty files will be created)
                        handle2 = Entrez.efetch(db=ncbi_db, id=list(record['IdList']), rettype="fasta", retmode="text")
                        all_records = SeqIO.parse(handle2, 'fasta')
                        count = SeqIO.write(all_records, out_file, "fasta-2line")

                        # check if data are really created & 
                        # loop through all asseccion numbers and write into searchterm , organism, and file accession number 
                        if os.path.isfile(out_file):
                            for records in SeqIO.parse(handle, 'genbank'):
                                f_assec.write(genus+ "\t"+  records.annotations['organism']+"\t" + records.id + "\t" + out_file  +"\t" + (";").join(records.annotations['taxonomy']) + "\n")
       
                    break
                except:
                    print("Something went wrong")
                    attempts += 1
                    f_log.write(genus+","+ new_search +"\n") 

exit()

File: 05_Check_NCBI.R
rm(list=ls())

source("00_Function_Library.R") # read file with functions
read_simple_ini("00_login_data.ini")
#####

library(rentrez)
library(stringr)

##################################################################################
# Aim: Download species from NCBI to create a Reference Database.
# Function:  Searches NCBI for nuclear 18S sequences of missing or related species, filters results, and downloads relevant sequences. 
##### The created folder and metadata will contain the data when the downloading was started.
######
# 1.) Species-level NCBI search
# 2.) Genus-level fallback search 
##################################################################################


 
### flags  ────────────────────────────────────────────────────────────────

FLAG_search_spec <- TRUE   # Step 1 – search by species
FLAG_search_gen  <- TRUE   # Step 2 – fall back to genus

### NCBI search parameters  ────────────────────────────────────────────────────────────────

search_suffix <- paste0(
  " AND (18S OR small ribosomal)",
  " NOT (chloroplast[filter] OR mitochondrion[filter])",
  " AND (350[SLEN] : 5000[SLEN])"
)
ncbi_db <- "nucleotide"

## input file paths ────────────────────────────────────────────────────────────────
Input_folder  <- file.path(path_to_output, "01_intermediate_results/")

file_missing            <- paste0(Input_folder, "4.1_Missing_Species.csv")
file_missing_but_genus  <- paste0(Input_folder, "4.2_Present_Species.csv")
file_missing_but_genus_new <- paste0(Input_folder,"4.2_Present_Species_completed_withNCBI.csv")
file.copy(file_missing_but_genus, file_missing_but_genus_new, overwrite = TRUE)

# ── entrez configuration ─────────────────────────────────────────────────────────────
TIMEOUT      <- 30          # seconds (used as httr / curl option where possible)

# Read email from config
set_entrez_key(NULL)        # clear any stale API key
options(ENTREZ_KEY = NULL)
entrez_email <- Entrez_email   # rentrez uses this automatically

# ── create new file paths ────────────────────────────────────────────────────────────────

#date <- format(Sys.Date(), "%d_%b_%Y")
date <- "20_Apr_2026"


NCBI_folder         <- file.path(path_to_output,paste0("02_NCBI_", date, "/"))
NCBI_results_folder <- file.path(path_to_output,paste0("02_NCBI_", date, "_results/"))
output_prefix       <- "05_Missing_PR2"

path_creation(c(NCBI_folder,NCBI_results_folder ))

file_still_missing   <- paste0(NCBI_results_folder, output_prefix,"_Species_and_missing_NCBI__",  date, ".csv")
file_missing_genera  <- paste0(NCBI_results_folder, output_prefix,"_incomplete_genera_despite_NCBI__", date, ".csv")
file_missing_genera2 <- paste0(NCBI_results_folder, output_prefix,"_MISSING_genera_despite_NCBI__", date, ".csv")
file_accession_info  <- paste0(NCBI_results_folder, output_prefix,"_Downloaded_species_NCBI_info__", date, ".tsv")
file_LOG             <- paste0(NCBI_results_folder, output_prefix,"LOG__", date, ".txt")
file_LOG2            <- paste0(NCBI_results_folder, output_prefix,"LOG__", date, "_successfull_genera.txt")

##############################################################################
# STEP 1 – Search and download by species
##############################################################################

if (FLAG_search_spec) {
  
  message("Start downloading species")
  
  con_missing   <- file(file_missing,            "r")
  con_still     <- file(file_still_missing,      "w")
  con_found     <- file(file_missing_but_genus_new, "a")
  con_log       <- file(file_LOG,                "a")
  con_assec     <- file(file_accession_info,     "w")
  
  lines <- readLines(con_missing)
  lines <- lines[-1]                    # drop header
  
  #line <- lines[2]
  for (line in lines) {
    line <- trimws(line)
    line <- gsub('"', '', line)
    parts   <- strsplit(line, ",")[[1]]
    genus   <- parts[1]
    species <- parts[2]
    
    message(species)
    new_search <- paste0(species, "[ORGANISM]", search_suffix)
    writeLines(new_search, con_log)
    
    success <- FALSE
    for (attempt in seq_len(3)) {
      message("  attempt ", attempt)
      result <- tryCatch({
        
        #tax <- entrez_search(db = "taxonomy", term = "Chironomus aprilinus")
        #tax$ids
        
        #new_search <- "Chironomus aprilinus[ORGANISM:noexp] AND (18S OR small ribosomal) NOT (chloroplast[filter] OR mitochondrion[filter]) AND (350[SLEN] : 5000[SLEN])"
        # --- search ---
        search_res <- entrez_search(db = ncbi_db, term = new_search,retmax = 10, use_history = FALSE)
        cat(search_res$QueryTranslation)
        
        search_res <- entrez_search(db = ncbi_db, term = new_search, retmax = 10, use_history = FALSE)
        summaries <- entrez_summary(db = ncbi_db, id = search_res$ids)
        
        # extract organism info
        sapply(summaries, function(x) x$organism)
        id_list <- search_res$ids
        
        if (length(id_list) == 0) {
          writeLines(line, con_still)
          success <- TRUE
          break
        }
        
        message("  downloading...")
        out_file <- paste0(NCBI_folder,gsub(" ", "_", species), "__NCBI.fasta")
        
        # --- fetch FASTA ---
        fasta_text <- entrez_fetch(db = ncbi_db, id = id_list,rettype = "fasta", retmode = "text")
        writeLines(fasta_text, out_file)
        
        # --- fetch GenBank ---
        if (file.exists(out_file) && file.info(out_file)$size > 0) {
          writeLines(paste0(genus, ",", species), con_found)
          
          gb_text <- entrez_fetch(db = ncbi_db, id = id_list,rettype = "genbank", retmode = "text")
          
          info_df <- parse_genbank_info(gb_text, species)
          info_df$file <- out_file
          write.table(info_df, con_assec,sep = "\t", col.names = FALSE, row.names = FALSE,quote = FALSE, append = TRUE)
        }
        
        success <- TRUE
        "ok"
        
      }, error = function(e) {
        msg <- paste0(line, "\tproblem: ", class(e)[1], " - ", conditionMessage(e))
        message("  ERROR: ", msg)
        writeLines(msg, con_log)
        Sys.sleep(2)
        e
      })
      
      if (isTRUE(success)) break
    }
  }
  
  close(con_missing)
  close(con_still)
  close(con_found)
  close(con_log)
  close(con_assec)
}



##############################################################################
# STEP 2 – Fall back to genus-level search
##############################################################################

if (FLAG_search_gen) {
  message("Start downloading species from the same genus")
  
  # -- 2a) Build genera_dict: genus -> vector of already-known species --------
  genera_dict <- list()
  lines_genus <- readLines(file_missing_but_genus_new)
  lines_genus <- lines_genus[-1]         # drop header
  
  for (line in lines_genus) {
    line  <- trimws(gsub('"', '', line))
    parts <- strsplit(line, ",")[[1]]
    g <- parts[1]; sp <- parts[2]
    if (g %in% names(genera_dict)) {
      genera_dict[[g]] <- c(genera_dict[[g]], sp)
    } else {
      genera_dict[[g]] <- sp
    }
  }
  
  # -- 2b) Build list of genera that still need searching --------------------
  genera_search <- character(0)
  if (file.exists(file_still_missing)) {
    lines_still <- readLines(file_still_missing)
    for (line in lines_still) {
      line  <- trimws(gsub('"', '', line))
      parts <- strsplit(line, ",")[[1]]
      g <- parts[1]
      if (!(g %in% genera_search)) genera_search <- c(genera_search, g)
    }
  }
  
  # -- 2c) Search & download by genus ----------------------------------------
  con_incomp <- file(file_missing_genera,  "w")
  con_miss   <- file(file_missing_genera2, "w")
  con_log    <- file(file_LOG,             "a")
  con_assec  <- file(file_accession_info,  "a")
  con_log2   <- file(file_LOG2,            "w")
  
  genera_search[genera_search %in% names(genera_dict)]
  genus <- "Acartia"
  for (genus in genera_search) {
    message("\n", genus)
    
    
    # Build search term excluding already-downloaded species
    if (genus %in% names(genera_dict)) {
      exclusions <- paste0(genera_dict[[genus]], "[ORGANISM]", collapse = " OR ")
      new_search <- paste0(genus, "[ORGANISM] NOT (", exclusions, ") ", search_suffix)
    } else {
      new_search <- paste0(genus, "[ORGANISM] ", search_suffix)
    }
    
    success <- FALSE
    for (attempt in seq_len(3)) {
      result <- tryCatch({
        
        # --- search ---
        search_res <- entrez_search(db = ncbi_db, term = new_search,
                                    retmax = 50, use_history = FALSE)
        id_list <- search_res$ids
        
        if (length(id_list) == 0) {
          message("Genus unknown: ", genus)
          if (genus %in% names(genera_dict)) {
            writeLines(paste0(genus, ",",paste(genera_dict[[genus]], collapse = "&"),new_search), con_miss)
          } else {
            writeLines(paste0(genus, ",", new_search), con_incomp)
          }
        } else {
          writeLines(paste0(genus, ",", new_search), con_log2)
          
          out_file <- paste0(NCBI_folder, genus, "__NCBI.fasta")
          
          # --- fetch FASTA ---
          fasta_text <- entrez_fetch(db = ncbi_db, id = id_list, rettype = "fasta", retmode = "text")
          writeLines(fasta_text, out_file)
          
          # --- fetch GenBank and write accession info ---
          if (file.exists(out_file) && file.info(out_file)$size > 0) {
            gb_text <- entrez_fetch(db = ncbi_db, id = id_list,rettype = "genbank", retmode = "text")
            
            info_df <- parse_genbank_info(gb_text, genus)
            info_df$file <- out_file
            write.table(info_df, con_assec,
                        sep = "\t", col.names = FALSE, row.names = FALSE,
                        quote = FALSE, append = TRUE)
          }
        }
        
        success <- TRUE
        "ok"
        
      }, error = function(e) {
        message("Something went wrong: ", conditionMessage(e))
        writeLines(paste0(genus, ",", new_search), con_log)
        Sys.sleep(2)
        e
      })
      
      if (isTRUE(success)) break
    }
  }
  
  close(con_incomp)
  close(con_miss)
  close(con_log)
  close(con_assec)
  close(con_log2)
}

message("Done.")
